src/coreclr/jit/lower.cpp

```diff
@@ -7844,6 +7860,126 @@ void Lowering::ContainCheckBitCast(GenTree* node)
+//
+void Lowering::LowerBlockStoreAsHelperCall(GenTreeBlk* blkNode)
+{
+    // We shouldn't be using helper calls for blocks on heap containing GC pointers.
+    // due to atomicity guarantees.
+    assert(!blkNode->IsZeroingGcPointersOnHeap());
+
+    LIR::Use use;
+    assert(!BlockRange().TryGetUse(blkNode, &use));
+
+    const bool isVolatile = blkNode->IsVolatile();
+
+    GenTree* dest = blkNode->Addr();
+    GenTree* data = blkNode->Data();
+    GenTree* size;
+
+    // Is it Memset ...
+    if (blkNode->OperIsInitBlkOp())
+    {
+        // Drop GT_INIT_VAL nodes
+        if (data->OperIsInitVal())
+        {
+            BlockRange().Remove(data);
+            data = data->gtGetOp1();
+        }
+    }
+    else
+    {
+        // ... or Memcpy?
+        assert(data->OperIs(GT_IND, GT_LCL_VAR, GT_LCL_FLD));
+
+        // Drop GT_IND nodes
+        if (data->OperIs(GT_IND))
+        {
+            BlockRange().Remove(data);
+            data = data->AsIndir()->Addr();
+        }
+
+        if (varTypeIsStruct(data))
+        {
+            // If it's a struct local, take its address
+            const unsigned lclNum = data->AsLclVarCommon()->GetLclNum();
+            GenTreeLclFld* addr   = comp->gtNewLclAddrNode(lclNum, data->AsLclVarCommon()->GetLclOffs(), TYP_BYREF);
+            BlockRange().InsertAfter(data, addr);
+            BlockRange().Remove(data);
+            data = addr;
+        }
+    }
+
+    if (blkNode->OperIs(GT_STORE_DYN_BLK))
+    {
+        // Size is not a constant
+        size = blkNode->AsStoreDynBlk()->gtDynamicSize;
+    }
+    else
+    {
+        // Size is a constant
+        size = comp->gtNewIconNode(blkNode->Size(), TYP_I_IMPL);
+        BlockRange().InsertBefore(data, size);
+    }
+
+    // A hacky way to safely call fgMorphTree in Lower
+    GenTree* destPlaceholder = comp->gtNewZeroConNode(dest->TypeGet());
+    GenTree* dataPlaceholder = comp->gtNewZeroConNode(data->TypeGet());
+    GenTree* sizePlaceholder = comp->gtNewZeroConNode(size->TypeGet());
+
+    CorInfoHelpFunc helper = blkNode->OperIsInitBlkOp() ? CORINFO_HELP_MEMSET : CORINFO_HELP_MEMCPY;
+    GenTreeCall* call = comp->gtNewHelperCallNode(helper, TYP_VOID, destPlaceholder, dataPlaceholder, sizePlaceholder);
+    comp->fgMorphTree(call);
+
+    LIR::Range range      = LIR::SeqTree(comp, call);
+    GenTree*   rangeStart = range.FirstNode();
+    GenTree*   rangeEnd   = range.LastNode();
+
+    BlockRange().InsertBefore(blkNode, std::move(range));
+    blkNode->gtBashToNOP();
+
+    LIR::Use destUse;
+    LIR::Use dataUse;
+    LIR::Use sizeUse;
+    BlockRange().TryGetUse(destPlaceholder, &destUse);
+    BlockRange().TryGetUse(dataPlaceholder, &dataUse);
+    BlockRange().TryGetUse(sizePlaceholder, &sizeUse);
+    destUse.ReplaceWith(dest);
+    dataUse.ReplaceWith(data);
+    sizeUse.ReplaceWith(size);
+    destPlaceholder->SetUnusedValue();
+    dataPlaceholder->SetUnusedValue();
+    sizePlaceholder->SetUnusedValue();
```

Comment:
Looks unnecessary since they're being removed below. I assume we can't just remove them here due to the `LowerRange` call?

Reply:
it works even without them yeah, but I just didn't want LIR to be in invalid state during LowerRange